src/Reports/Full.php

```diff
@@ -115,36 +115,49 @@ public function generateFileReport($report, File $phpcsFile, $showSources=false,
         // The maximum amount of space an error message can use.
         $maxErrorSpace = ($width - $paddingLength - 1);
 
+        $beforeMsg = '';
+        $afterMsg  = '';
+        if ($showSources === true) {
+            $beforeMsg = "\033[1m";
+            $afterMsg  = "\033[0m";
+        }
+
         foreach ($report['messages'] as $line => $lineErrors) {
             foreach ($lineErrors as $column => $colErrors) {
                 foreach ($colErrors as $error) {
-                    $message  = $error['message'];
-                    $msgLines = [$message];
-                    if (strpos($message, "\n") !== false) {
-                        $msgLines = explode("\n", $message);
-                    }
+                    $errorMsg = wordwrap(
+                        $error['message'],
+                        $maxErrorSpace
+                    );
 
-                    $errorMsg = '';
-                    $lastLine = (count($msgLines) - 1);
-                    foreach ($msgLines as $k => $msgLine) {
-                        if ($k === 0) {
-                            if ($showSources === true) {
-                                $errorMsg .= "\033[1m";
-                            }
-                        } else {
-                            $errorMsg .= PHP_EOL.$paddingLine2;
-                        }
+                    // Add the padding _after_ the wordwrap as the message itself may contain line breaks
+                    // and those lines will also need to receive padding.
+                    $errorMsg = str_replace("\n", $afterMsg.PHP_EOL.$paddingLine2.$beforeMsg, $errorMsg);
+                    $errorMsg = $beforeMsg.$errorMsg.$afterMsg;
 
-                        if ($k === $lastLine && $showSources === true) {
-                            $msgLine .= "\033[0m".' ('.$error['source'].')';
+                    if ($showSources === true) {
+                        $lastMsg          = $errorMsg;
+                        $startPosLastLine = strrpos($errorMsg, $paddingLine2.$beforeMsg);
+                        if ($startPosLastLine !== false) {
+                            // Message is multiline.
+                            $lastMsg = substr($errorMsg, ($startPosLastLine + strlen($paddingLine2.$beforeMsg)));
                         }
```

Comment:
It's pretty unlikely that some message will contain something like "   |         |     ", but perhaps we should still include the leading PHP_EOL here too for good measure.

On the other hand, stripping out the `$beforeMsg` on line 143 leads to inconsistency below on line 153: if `$startPosLastLine` was `false`, 8 is correct there because the string contains both `$beforeMsg` and `$afterMsg`, but if it wasn't then we stripped off the `$beforeMsg` so only 4 are needed.
```suggestion
                        $startPosLastLine = strrpos($errorMsg, PHP_EOL.$paddingLine2.$beforeMsg);
                        if ($startPosLastLine !== false) {
                            // Message is multiline. Take just the last line (starting at the `$beforeMsg`).
                            $lastMsg = substr($errorMsg, ($startPosLastLine + strlen(PHP_EOL.$paddingLine2)));
                        }
```

Seeing the result of the inconsistency is kind of tough. Adding `$phpcsFile->addWarning( str_repeat( 'x ', 100 ) . 'x', 0, 'Test' );` in some sniff so single characters wrap can show it.

Reply:
Nice catch! And yes, I agree with the bug analysis and the proposed fix. Updated.,.